src/cryptsetup/cryptsetup.c

```diff
@@ -2497,27 +2557,26 @@ static int run(int argc, char *argv[]) {
                         /* Key not correct? Let's try again, but let's invalidate one of the passed fields,
                          * so that we fallback to the next best thing. */
 
-                        if (arg_tpm2_device || arg_tpm2_device_auto) {
+                        if (token_type == TOKEN_TPM2) {
```

Comment:
of course, changing this line like this is ok, but I'd probably not do it, simply because if you leave this as is the relationship of the check and the invalidation below is really clear: we reset fields if they are set, and try again.

Reply:
I see your point, but then we would have to make sure that the order of invalidation matches the order in `determine_token_type` and the whole point of enuming the token types was to centralize the token ordering.